mirrord/auth/src/credentials.rs

```diff
@@ -70,6 +70,119 @@ impl AsRef<Certificate> for Credentials {
     }
 }
 
+/// Gives some more meaning to a `License`'s [`Certificate`] expiration date, so we don't have
+/// to manually check dates to see if the licese is still good or not.
+///
+/// You may use the [`From`] implementation to build a [`LicenseValidity`] from a [`NaiveDate`],
+/// which is what we store in the `LicenseInfoOwned` type.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LicenseValidity {
+    /// The license is still far away from expiring (see `LicenseValidity::new`).
+    ///
+    /// Holds the expiration date of the license.
+    Good(DateTime<Utc>),
+
+    /// License has expired.
+    ///
+    /// Holds the expiration date of the license.
+    Expired(DateTime<Utc>),
+}
+
+impl LicenseValidity {
+    /// Builds the [`LicenseValidity`]. You should use either this or the [`From`] implementation.
+    ///
+    /// We take `now` as a parameter here so we can have [`LicenseValidity`] with fake values.
+    pub fn new(expiration_date: DateTime<Utc>, now: DateTime<Utc>) -> Self {
+        if now > expiration_date {
+            Self::Expired(expiration_date)
+        } else {
+            Self::Good(expiration_date)
+        }
+    }
+
+    /// Checks if this [`LicenseValidity`] represents a date that is close to expiring.
+    ///
+    /// We consider close if the `License` is 2 days away from its `expiration_date`.
+    pub fn close_to_expiring(&self) -> Option<i64> {
```

Comment:
Instead of implementing this as a method, I suggest changing the enum to have 3 variants: `Good`, `CloseToExpire` and `Expired`. Rationale:
1. This method checks something, so it should probably return a boolean. 
2. An expired license is technically not close to be expired, so this method should return false.
3. Which would make it not practical :v

Reply:
Guess I should read all the comments before addressing them lol.

Had this a 3-way enum, but on the operator side `CloseToExpiring` is meaningless, so everywhere I was `match`ing on it and doing the same thing as `Good`. So I removed in favor of the current 2-way enum.

I don't want to return a boolean here, so we can return how many days are left before expiration, to give a nice warning to the user (and not have to extract the expiration date and do the math on the warning call-site).

The main idea here is making it easy to get how many days are left, so we can display it.

Reply:
Maybe something like this would be simplest
```rust
struct LicenseValidity(DateTime<Utc>);

impl LicenseValidity {
    fn is_good(&self) -> bool { ... };
    fn is_expired(&self) -> bool { ... };
    fn days_until_expiration(&self) -> Option<u32> { ... };
}
```
What do you think?

Reply:
Hmm if we go this route I think we could drop the type and just have these methods on the license, right?

Reply:
Nope, not on the `License`, as we also need them for `LicenseInfoOwned`, so I guess they would have to be free functions.

Reply:
This is also an option:
```rust
trait LicenseValidity {
    fn is_good(&self) -> bool { ... };
    fn days_until_expiration(&self) -> Option<u32> { ... };
}

impl LicenseValidity for DateTime<Utc> { ... }
```

Reply:
I like this suggestion the most so far, but I'm not a fan of `bools` and prefer types.

Left both implementations for you to compare, on the operator side the change would be instead of `match license.validity` we do `if license.is_good() { ... } else { ... }`.

Leaving the final call on this to you.

Reply:
I like `LicenseValidityV2` more. Sorry for being a pain in the lower back :sweat_smile: